src/poprox_recommender/default.py

```diff
@@ -1,91 +1,55 @@
         user_embedder = UserEmbedder(model.model, model.device)
         article_scorer = ArticleScorer(model.model)
 
-        candidate_articles = article_embedder(candidate_articles)
-        clicked_articles = article_embedder(clicked_articles)
-
-        interest_profile = user_embedder(clicked_articles, interest_profile)
-        candidate_articles = article_scorer(candidate_articles, interest_profile)
-
         if diversify == "mmr":
             diversifier = MMRDiversifier(algo_params, num_slots)
-            recs = diversifier(candidate_articles)
-
         elif diversify == "pfar":
             diversifier = PFARDiversifier(algo_params, num_slots)
-            recs = diversifier(candidate_articles, interest_profile)
-
-        recommendations[account_id] = recs.articles
-    else:
-        recommendations[account_id] = select_by_topic(
-            candidate_articles.articles,
-            interest_profile,
-            num_slots,
-        )
-
-    return recommendations
-
-
-def select_by_topic(todays_articles: list[Article], interest_profile: InterestProfile, num_slots: int):
-    # Preference values from onboarding are 1-indexed, where 1 means "absolutely no interest."
-    # We might want to normalize them to 0-indexed somewhere upstream, but in the mean time
-    # this is one of the simpler ways to filter out topics people aren't interested in from
-    # their early newsletters
-    profile_topics = {
-        interest.entity_name for interest in interest_profile.onboarding_topics if interest.preference > 1
-    }
-
-    other_articles = []
-    topical_articles = []
-    for article in todays_articles:
-        article_topics = {mention.entity.name for mention in article.mentions}
-        if len(profile_topics.intersection(article_topics)) > 0:
-            topical_articles.append(article)
-        else:
-            other_articles.append(article)
 
-    if len(topical_articles) >= num_slots:
-        return random.sample(topical_articles, num_slots)
+        pipeline = RecommendationPipeline(name=diversify)
+        pipeline.add(article_embedder, inputs=["candidate"], output="candidate")
+        pipeline.add(article_embedder, inputs=["clicked"], output="clicked")
+        pipeline.add(user_embedder, inputs=["clicked", "profile"], output="profile")
+        pipeline.add(article_scorer, inputs=["candidate", "profile"], output="candidate")
+        pipeline.add(diversifier, inputs=["candidate", "profile"], output="recs")
     else:
-        return random.sample(topical_articles, len(topical_articles)) + random.sample(
-            other_articles, num_slots - len(topical_articles)
-        )
+        topic_filter = TopicFilter()
+        sampler = UniformSampler(num_slots=num_slots)
+
+        pipeline = RecommendationPipeline(name="random_topical")
+        pipeline.add(topic_filter, inputs=["candidate", "profile"], output="topical")
+        pipeline.add(sampler, inputs=["topical", "candidate"], output="recs")
```

Comment:
In the future, I'd like to make these recommendation pipelines serializable to/from YAML (or similar.) Then the handler could read all the recommender definition files present in a directory, load the pipelines into a dictionary by name, and be prepared to serve any of them based on a query string parameter. That would give us an easy way to map from the names of the recommenders in the experiment manifest to the actual pipeline that matches it, and allow people to define recommenders without writing any Python (as long as they use existing components we provide.)

Reply:
Given the amount of machinery needed to handle initialization, checkpointing, training, etc., I would probably put this as a relatively low priority for poprox.

Reply:
I suspect we see this differently because you're thinking of a pipeline that handles all the work required to prepare models and I'm not, particularly because support for all that machinery across multiple frameworks (TF, Torch, JAX, LensKit, etc) seems fairly impossible given our constrained resources.

We're going to have a default model (or several) that are pre-trained and available to experimenters, and I'm told that many people may want to use those models in recommenders that only modify things like presentation logic only. So, with this idea I'm responding to earlier feedback that forking a repo and modifying code is going to be much harder for e.g. social scientists than hitting an endpoint. (You may remember my perspective to the contrary was described as an example of my "developer brain" not having empathy for non-developers.)

Reply:
"Here's a library of pre-fab semi-configurable recommender components, here's a text-editable file format that defines executable recommenders, here's how to use it to build experiment recommenders" seems like the kind of thing that might add value for our target audience (and might help with our "what exactly do we document?" conundrum too)

Reply:
Makes sense. If we're very tightly-scoped I can see it working.

The challenge I see — setting aside training and generating checkpoints — is the number of different ways components can be instantiated. Already we have different sources of component data:

- our own checkpoint files in safetensors
- pretrained models from huggingface
- python components configured with parameters

Now, we could simplify (1) and (2) into additional parameters, the paths to the model data. That might make it tractable.

Reply:
Yeah, as long as the components take basic Python types as constructor parameters (like model names or string paths as you suggest), it should be easy enough to serialize the params and reconstruct the objects.

That may influence how we design components though—for example, we might have `SafeTensorsScoringModel` and `HuggingFaceScoringModel` components instead of one that does both. That doesn't necessarily mean that we need to write a component for every framework as long the ones we do provide lay down a pattern that people can follow.